src/representative_subset_calculator/streaming/candidate_consumer.h

```diff
@@ -159,4 +158,156 @@ class SeiveCandidateConsumer : public CandidateConsumer {
 
         return highestMarginal;
     }
+};
+
+
+class ThreeSeiveCandidateConsumer : public CandidateConsumer {
+    private:
+    const unsigned int numberOfSenders;
+    const unsigned int k;
+    const double epsilon;
+    unsigned int T; // hyperparameter -- tranfer contents to next bucket after T unsuccesful inserts
+    unsigned int t; // counts till T
+    unsigned int currentBucketIndex;
+
+    std::vector<std::unique_ptr<CandidateSeed>> seedStorage;
+    std::unordered_set<unsigned int> seenFirstElement;
+    std::unique_ptr<std::vector<ThresholdBucket>> buckets = std::make_unique<std::vector<ThresholdBucket>>();
+    std::vector<double> firstMarginals;
+    std::unordered_set<unsigned int> firstGlobalRows;
+
+    public:
+    ThreeSeiveCandidateConsumer(
+        const unsigned int numberOfSenders, 
+        const unsigned int k, 
+        const double epsilon,
+        unsigned int T
+    ) : 
+        numberOfSenders(numberOfSenders),
+        k(k),
+        epsilon(epsilon),
+        T(T),
+        t(0),
+        currentBucketIndex(0)
+    {}
+
+    void accept(SynchronousQueue<std::unique_ptr<CandidateSeed>> &seedQueue, Timers &timers) {
+        if (!this->bucketsInitialized()) {
+            timers.initBucketsTimer.startTimer();
+            this->tryToGetFirstMarginals(seedQueue);
+            timers.initBucketsTimer.stopTimer();
+        } 
+        
+        if (this->bucketsInitialized()) {
+            timers.insertSeedsTimer.startTimer();
+            this->processQueue(seedQueue);
+            timers.insertSeedsTimer.stopTimer();
+        }
+    }
+
+    std::unique_ptr<Subset> getBestSolutionDestroyConsumer() {
+
+        return (*(this->buckets))[this->currentBucketIndex].returnSolutionDestroyBucket();
+    }
+
+    private:
+    bool bucketsInitialized() const {
+        return buckets->size() != 0;
+    }
+
+    unsigned int getNumberOfBuckets(const unsigned int k, const double epsilon) {
+        int numBuckets = (int)(0.5 + [](double val, double base) {
+            return log2(val) / log2(base);
+        }((double)k, (1+this->epsilon))) + 1;
+
+        return numBuckets;
+    }
+
+    double getDeltaZero() {
+        double highestMarginal = 0;
+        for (const double m : this->firstMarginals) {
+            highestMarginal = std::max(highestMarginal, m);
+        }
+
+        return highestMarginal;
+    }
+
+    void initBuckets(SynchronousQueue<std::unique_ptr<CandidateSeed>> &seedQueue) {
+        const double deltaZero = this->getDeltaZero();
+        const unsigned int numBuckets = this->getNumberOfBuckets(this->k, this->epsilon);
+
+        std::cout << "number of buckets " << numBuckets << " with deltaZero of " << deltaZero << std::endl;
+        // TODO:: why that extra 2 in (delta_zero / 2k). Check later
+        double threshold = ((double)deltaZero / (double)( 2 * k )) * (double)std::pow(1 + epsilon, numBuckets - 1);
+        this->buckets->push_back(ThresholdBucket(threshold, k));
```

Comment:
Looks like we're only maintaining one bucket at a time. `this->buckets` should be `this->bucket`